<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkRP0321ENSkillsNetwork25371262-2022-01-01">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-RP0101EN-Coursera/v2/M1_R_Basics/images/IDSNlogo.png" width="200" align="center">
</a>


<h1>Data Wrangling with dplyr</h1>

Estimated time needed: **60** minutes


## Lab Overview:

In this lab, you will focus on wrangling the Seoul bike-sharing demand historical dataset. This is the core dataset to build a predictive model later.

It contains the following columns:

*   `DATE` : Year-month-day
*   `RENTED BIKE COUNT`- Count of bikes rented at each hour
*   `HOUR`- Hour of he day
*   `TEMPERATURE` - Temperature in Celsius
*   `HUMIDITY` - Unit is `%`
*   `WINDSPEED` - Unit is `m/s`
*   `VISIBILITY` - Multiplied by 10m
*   `DEW POINT TEMERATURE` - The temperature to which the air would have to cool down in order to reach saturation, unit is Celsius
*   `SOLAR RADIATION` - MJ/m2
*   `RAINFALL` - mm
*   `SNOWFALL` - cm
*   `SEASONS` - Winter, Spring, Summer, Autumn
*   `HOLIDAY` - Holiday/No holiday
*   `FUNCTIONAL DAY` - NoFunc(Non Functional Hours), Fun(Functional hours)

For this dataset, you will be asked to use `tidyverse` to perform the following data wrangling tasks:

*   **TASK: Detect and handle missing values**
*   **TASK: Create indicator (dummy) variables for categorical variables**
*   **TASK: Normalize data**

Let's start!


First import the necessary library for this data wrangling task:


In [68]:
# Check if you need to install the `tidyverse` library
require("tidyverse")
library(tidyverse)

Then load the bike-sharing system data from the csv processed in the previous lab:


In [69]:
bike_sharing_df <- read_csv("raw_seoul_bike_sharing.csv")

Rows: 8760 Columns: 14
-- Column specification --------------------------------------------------------
Delimiter: ","
chr  (4): Date, SEASONS, HOLIDAY, FUNCTIONING_DAY
dbl (10): RENTED_BIKE_COUNT, Hour, TEMPERATURE, HUMIDITY, WIND_SPEED, Visibi...

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.


In [70]:
# Or you may read it from here again
# url <- "https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-RP0321EN-SkillsNetwork/labs/datasets/raw_seoul_bike_sharing.csv"
# Notice some column names in the raw datasets are not standalized if you haven't done them properly in the previous lab

First take a quick look at the dataset:


In [71]:
summary(bike_sharing_df)
dim(bike_sharing_df)

     Date           RENTED_BIKE_COUNT      Hour        TEMPERATURE    
 Length:8760        Min.   :   2.0    Min.   : 0.00   Min.   :-17.80  
 Class :character   1st Qu.: 214.0    1st Qu.: 5.75   1st Qu.:  3.40  
 Mode  :character   Median : 542.0    Median :11.50   Median : 13.70  
                    Mean   : 729.2    Mean   :11.50   Mean   : 12.87  
                    3rd Qu.:1084.0    3rd Qu.:17.25   3rd Qu.: 22.50  
                    Max.   :3556.0    Max.   :23.00   Max.   : 39.40  
                    NA's   :295                       NA's   :11      
    HUMIDITY       WIND_SPEED      Visibility   DEW_POINT_TEMPERATURE
 Min.   : 0.00   Min.   :0.000   Min.   :  27   Min.   :-30.600      
 1st Qu.:42.00   1st Qu.:0.900   1st Qu.: 940   1st Qu.: -4.700      
 Median :57.00   Median :1.500   Median :1698   Median :  5.100      
 Mean   :58.23   Mean   :1.725   Mean   :1437   Mean   :  4.074      
 3rd Qu.:74.00   3rd Qu.:2.300   3rd Qu.:2000   3rd Qu.: 14.800      
 Max.   :98.

[1] 8760   14

From the summary, we can observe that:

Columns `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL` are numerical variables/columns and require normalization. Moreover, `RENTED_BIKE_COUNT` and `TEMPERATURE` have some missing values (NA's) that need to be handled properly.

`SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY` are categorical variables which need to be converted into indicator columns or dummy variables.
Also, `HOUR` is read as a numerical variable but it is in fact a categorical variable with levels ranging from 0 to 23.

Now that you have some basic ideas about how to process this bike-sharing demand dataset, let's start working on it!


# TASK: Detect and handle missing values


The `RENTED_BIKE_COUNT` column has about 295 missing values, and `TEMPERATURE` has about 11 missing values. Those missing values could be caused by not being recorded, or from malfunctioning bike-sharing systems or weather sensor networks. In any cases, the identified missing values have to be properly handled.


Let's first handle missing values in `RENTED_BIKE_COUNT` column:


Considering `RENTED_BIKE_COUNT` is the response variable/dependent variable, i.e., we want to predict the `RENTED_BIKE_COUNT` using other predictor/independent variables later, and we normally can not allow missing values for the response variable, so missing values for response variable must be either dropped or imputed properly.

We can see that `RENTED_BIKE_COUNT` only has about 3% missing values (295 / 8760). As such, you can safely drop any rows whose `RENTED_BIKE_COUNT` has missing values.


*TODO:* Drop rows with missing values in the `RENTED_BIKE_COUNT` column


In [72]:
# Drop rows with `RENTED_BIKE_COUNT` column == NA
cleaned_bike_sharing<-bike_sharing_df %>%
drop_na(RENTED_BIKE_COUNT)

In [73]:
# Print the dataset dimension again after those rows are dropped
dim(cleaned_bike_sharing)

[1] 8465   14

Now that you have handled  missing values in the `RENTED_BIKE_COUNT` variable, let's continue processing missing values for the `TEMPERATURE` column.


Unlike the `RENTED_BIKE_COUNT` variable, `TEMPERATURE` is not a response variable. However, it is still an important predictor variable - as you could imagine, there may be a positve correlation between `TEMPERATURE` and `RENTED_BIKE_COUNT`. For example, in winter time with lower temperatures, people may not want to ride a bike, while in summer with nicer weather, they are more likely to rent a bike.


How do we handle missing values for `TEMPERATURE`? We could simply remove the rows but it's better to impute them because `TEMPERATURE` should be relatively easy and reliable to estimate statistically.


Let's first take a look at the missing values in the TEMPERATURE column.


In [126]:
cleaned_bike_sharing %>% 
                filter(is.na(TEMPERATURE))
head(cleaned_bike_sharing)

Date       RENTED_BIKE_COUNT Hour TEMPERATURE HUMIDITY WIND_SPEED Visibility
1  07/06/2018 3221              18   NA          57       2.7        1217      
2  12/06/2018 1246              14   NA          45       2.2        1961      
3  13/06/2018 2664              17   NA          57       3.3         919      
4  17/06/2018 2330              17   NA          58       3.3         865      
5  20/06/2018 2741              19   NA          61       2.7        1236      
6  30/06/2018 1144              13   NA          87       1.7         390      
7  05/07/2018  827              10   NA          75       1.1        1028      
8  11/07/2018  634               9   NA          96       0.6         450      
9  12/07/2018  593               6   NA          93       1.1         852      
10 21/07/2018  347               4   NA          77       1.2        1203      
11 21/08/2018 1277              23   NA          75       0.1        1892      
   DEW_POINT_TEMPERATURE SOLAR_RADIATION RAINFALL Snowfall SEASONS HOLIDAY   
1  16.4                  0.96            0.0      0        Summer  No Holiday
2  12.7                  1.39            0.0      0        Summer  No Holiday
3  16.4                  0.87            0.0      0        Summer  No Holiday
4  16.7                  0.66            0.0      0        Summer  No Holiday
5  17.5                  0.60            0.0      0        Summer  No Holiday
6  23.2                  0.71            3.5      0        Summer  No Holiday
7  20.8                  1.22            0.0      0        Summer  No Holiday
8  24.9                  0.41            0.0      0        Summer  No Holiday
9  24.3                  0.01            0.0      0        Summer  No Holiday
10 21.2                  0.00            0.0      0        Summer  No Holiday
11 20.8                  0.00            0.0      0        Summer  No Holiday
   FUNCTIONING_DAY
1  Yes            
2  Yes            
3  Yes            
4  Yes            
5  Yes            
6  Yes            
7  Yes            
8  Yes            
9  Yes            
10 Yes            
11 Yes

Date       RENTED_BIKE_COUNT Hour TEMPERATURE HUMIDITY WIND_SPEED Visibility
1 01/12/2017 254               0    -5.2        37       2.2        2000      
2 01/12/2017 204               1    -5.5        38       0.8        2000      
3 01/12/2017 173               2    -6.0        39       1.0        2000      
4 01/12/2017 107               3    -6.2        40       0.9        2000      
5 01/12/2017  78               4    -6.0        36       2.3        2000      
6 01/12/2017 100               5    -6.4        37       1.5        2000      
  DEW_POINT_TEMPERATURE SOLAR_RADIATION RAINFALL Snowfall SEASONS HOLIDAY   
1 -17.6                 0               0        0        Winter  No Holiday
2 -17.6                 0               0        0        Winter  No Holiday
3 -17.7                 0               0        0        Winter  No Holiday
4 -17.6                 0               0        0        Winter  No Holiday
5 -18.6                 0               0        0        Winter  No Holiday
6 -18.7                 0               0        0        Winter  No Holiday
  FUNCTIONING_DAY
1 Yes            
2 Yes            
3 Yes            
4 Yes            
5 Yes            
6 Yes

It seems that all of the missing values for `TEMPERATURE` are found in `SEASONS == Summer`, so it is reasonable to impute those missing values with the summer average temperature.


*TODO:* Impute missing values for the TEMPERATURE column using its mean value.


In [75]:
# Calculate the summer average temperature
filtered_df<-cleaned_bike_sharing %>% filter(SEASONS == "Summer")
summer_temperature<-mean(filtered_df[["TEMPERATURE"]],na.rm=TRUE)


In [76]:
# Impute missing values for TEMPERATURE column with summer average temperature
cleaner_bike_df<-cleaned_bike_sharing %>% replace_na(list(TEMPERATURE= 26.5))
convert<-toupper(colnames(cleaner_bike_df))
    colnames(cleaner_bike_df)<- convert

In [77]:
# Print the summary of the dataset again to make sure no missing values in all columns
summary(cleaner_bike_df)
glimpse(cleaner_bike_df)


     DATE           RENTED_BIKE_COUNT      HOUR        TEMPERATURE    
 Length:8465        Min.   :   2.0    Min.   : 0.00   Min.   :-17.80  
 Class :character   1st Qu.: 214.0    1st Qu.: 6.00   1st Qu.:  3.00  
 Mode  :character   Median : 542.0    Median :12.00   Median : 13.50  
                    Mean   : 729.2    Mean   :11.51   Mean   : 12.77  
                    3rd Qu.:1084.0    3rd Qu.:18.00   3rd Qu.: 22.70  
                    Max.   :3556.0    Max.   :23.00   Max.   : 39.40  
    HUMIDITY       WIND_SPEED      VISIBILITY   DEW_POINT_TEMPERATURE
 Min.   : 0.00   Min.   :0.000   Min.   :  27   Min.   :-30.600      
 1st Qu.:42.00   1st Qu.:0.900   1st Qu.: 935   1st Qu.: -5.100      
 Median :57.00   Median :1.500   Median :1690   Median :  4.700      
 Mean   :58.15   Mean   :1.726   Mean   :1434   Mean   :  3.945      
 3rd Qu.:74.00   3rd Qu.:2.300   3rd Qu.:2000   3rd Qu.: 15.200      
 Max.   :98.00   Max.   :7.400   Max.   :2000   Max.   : 27.200      
 SOLAR_RADIAT

Rows: 8,465
Columns: 14
$ DATE                  <chr> "01/12/2017", "01/12/2017", "01/12/2017", "01/12~
$ RENTED_BIKE_COUNT     <dbl> 254, 204, 173, 107, 78, 100, 181, 460, 930, 490,~
$ HOUR                  <dbl> 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14~
$ TEMPERATURE           <dbl> -5.2, -5.5, -6.0, -6.2, -6.0, -6.4, -6.6, -7.4, ~
$ HUMIDITY              <dbl> 37, 38, 39, 40, 36, 37, 35, 38, 37, 27, 24, 21, ~
$ WIND_SPEED            <dbl> 2.2, 0.8, 1.0, 0.9, 2.3, 1.5, 1.3, 0.9, 1.1, 0.5~
$ VISIBILITY            <dbl> 2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, ~
$ DEW_POINT_TEMPERATURE <dbl> -17.6, -17.6, -17.7, -17.6, -18.6, -18.7, -19.5,~
$ SOLAR_RADIATION       <dbl> 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, ~
$ RAINFALL              <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ~
$ SNOWFALL              <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ~
$ SEASONS               <chr> "Winter", "Winter", "Winter", "Winter", "Winter"~
$ HOLIDAY       

In [78]:
# Save the dataset as `seoul_bike_sharing.csv`
write.csv(cleaner_bike_df,"seoul_bike_sharing.csv",row.names=FALSE)

# TASK: Create indicator (dummy) variables for categorical variables


Regression models can not process categorical variables directly, thus we need to convert them into indicator variables.


In the bike-sharing demand dataset, `SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY` are categorical variables.
Also, `HOUR` is read as a numerical variable but it is in fact a categorical variable with levels ranged from 0 to 23.


*TODO:* Convert `HOUR` column from numeric into character first:


In [107]:
# Using mutate() function to convert HOUR column into character type
even_cleaner_df<- cleaner_bike_df %>% mutate(HOUR = as.character(HOUR))
glimpse(even_cleaner_df)

Rows: 8,465
Columns: 14
$ DATE                  <chr> "01/12/2017", "01/12/2017", "01/12/2017", "01/12~
$ RENTED_BIKE_COUNT     <dbl> 254, 204, 173, 107, 78, 100, 181, 460, 930, 490,~
$ HOUR                  <chr> "0", "1", "2", "3", "4", "5", "6", "7", "8", "9"~
$ TEMPERATURE           <dbl> -5.2, -5.5, -6.0, -6.2, -6.0, -6.4, -6.6, -7.4, ~
$ HUMIDITY              <dbl> 37, 38, 39, 40, 36, 37, 35, 38, 37, 27, 24, 21, ~
$ WIND_SPEED            <dbl> 2.2, 0.8, 1.0, 0.9, 2.3, 1.5, 1.3, 0.9, 1.1, 0.5~
$ VISIBILITY            <dbl> 2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, ~
$ DEW_POINT_TEMPERATURE <dbl> -17.6, -17.6, -17.7, -17.6, -18.6, -18.7, -19.5,~
$ SOLAR_RADIATION       <dbl> 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, ~
$ RAINFALL              <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ~
$ SNOWFALL              <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ~
$ SEASONS               <chr> "Winter", "Winter", "Winter", "Winter", "Winter"~
$ HOLIDAY       

`SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY`,  `HOUR` are all character columns now and are ready to be converted into indicator variables.

For example, `SEASONS` has four categorical values: `Spring`, `Summer`, `Autumn`, `Winter`. We thus need to create four indicator/dummy variables `Spring`, `Summer`, `Autumn`, and `Winter` which only have the value 0 or 1.

So, given a data entry with the value `Spring` in the `SEASONS` column, the values for the four new columns `Spring`, `Summer`, `Autumn`, and `Winter` will be set to 1 for `Spring` and 0 for the others:

| Spring | Summer | Autumn | Winter |
| ------ | ------ | ------ | ------ |
| 1      | 0      | 0      | 0      |


*TODO:* Convert `SEASONS`, `HOLIDAY`, `FUNCTIONING_DAY`, and `HOUR` columns into indicator columns.

Note that if `FUNCTIONING_DAY` only contains one categorical value after missing values removal, then you don't need to convert it to an indicator column.


In [111]:
# Convert SEASONS, HOLIDAY, FUNCTIONING_DAY, and HOUR columns into indicator columns.
df1<-even_cleaner_df%>%
mutate(dummy= 1) %>%
spread(key=SEASONS,
        value= 1,
         fill= 0)
df2<-df1 %>%
 mutate(dummy= 1) 
spread(key=HOLIDAY,
        value= 1,
         fill= 0)
mutate(dummy= 1) 
dummy_data<-df2  %>%
spread(key=HOUR,
        value= 1,
         fill= 0)

Rows: 8,465
Columns: 17
$ HOUR                  <chr> "0", "1", "2", "3", "4", "5", "6", "7", "8", "9"~
$ TEMPERATURE           <dbl> -5.2, -5.5, -6.0, -6.2, -6.0, -6.4, -6.6, -7.4, ~
$ HUMIDITY              <dbl> 37, 38, 39, 40, 36, 37, 35, 38, 37, 27, 24, 21, ~
$ WIND_SPEED            <dbl> 2.2, 0.8, 1.0, 0.9, 2.3, 1.5, 1.3, 0.9, 1.1, 0.5~
$ VISIBILITY            <dbl> 2000, 2000, 2000, 2000, 2000, 2000, 2000, 2000, ~
$ DEW_POINT_TEMPERATURE <dbl> -17.6, -17.6, -17.7, -17.6, -18.6, -18.7, -19.5,~
$ SOLAR_RADIATION       <dbl> 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, 0.00, ~
$ RAINFALL              <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ~
$ SNOWFALL              <dbl> 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ~
$ FUNCTIONING_DAY       <chr> "Yes", "Yes", "Yes", "Yes", "Yes", "Yes", "Yes",~
$ dummy                 <dbl> 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ~
$ Autumn                <chr> "0", "0", "0", "0", "0", "0", "0", "0", "0", "0"~
$ Spring        

In [112]:
# Print the dataset summary again to make sure the indicator columns are created properly
summary(dummy_data)

  TEMPERATURE        HUMIDITY       WIND_SPEED      VISIBILITY  
 Min.   :-17.80   Min.   : 0.00   Min.   :0.000   Min.   :  27  
 1st Qu.:  3.00   1st Qu.:42.00   1st Qu.:0.900   1st Qu.: 935  
 Median : 13.50   Median :57.00   Median :1.500   Median :1690  
 Mean   : 12.77   Mean   :58.15   Mean   :1.726   Mean   :1434  
 3rd Qu.: 22.70   3rd Qu.:74.00   3rd Qu.:2.300   3rd Qu.:2000  
 Max.   : 39.40   Max.   :98.00   Max.   :7.400   Max.   :2000  
 DEW_POINT_TEMPERATURE SOLAR_RADIATION     RAINFALL          SNOWFALL      
 Min.   :-30.600       Min.   :0.0000   Min.   : 0.0000   Min.   :0.00000  
 1st Qu.: -5.100       1st Qu.:0.0000   1st Qu.: 0.0000   1st Qu.:0.00000  
 Median :  4.700       Median :0.0100   Median : 0.0000   Median :0.00000  
 Mean   :  3.945       Mean   :0.5679   Mean   : 0.1491   Mean   :0.07769  
 3rd Qu.: 15.200       3rd Qu.:0.9300   3rd Qu.: 0.0000   3rd Qu.:0.00000  
 Max.   : 27.200       Max.   :3.5200   Max.   :35.0000   Max.   :8.80000  
 FUNCTIONING_

In [90]:
# Save the dataset as `seoul_bike_sharing_converted.csv`
# write_csv(dataframe, "seoul_bike_sharing_converted.csv")
write_csv(dummy_data,"seoul_bike_sharing_converted.csv")

# TASK: Normalize data


Columns `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL` are numerical variables/columns with different value units and range. Columns with large values may adversely influence (bias) the predictive models and degrade model accuracy. Thus, we need to perform normalization on these numeric columns to transfer them into a similar range.


In this project, you are asked to use Min-max normalization:

**Min-max** rescales each value in a column by first subtracting the minimum value of the column from each value, and then divides the result by the difference between the maximum and minimum values of the column. So the column gets re-scaled such that the minimum becomes 0 and the maximum becomes 1.

$$x\_{new} = \frac{x\_{old} - x\_{min}}{x\_{max} - x\_{min}}$$


*TODO:* Apply min-max normalization on `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL`


In [133]:
# Use the `mutate()` function to apply min-max normalization on columns 
min_max<-function(x){
    return((x-min(x))/(max(x)-min(x)))
    }
normalized_dummy<-dummy_data %>% mutate(    TEMPERATURE=min_max(TEMPERATURE),
                          HUMIDITY=min_max(HUMIDITY),
                          WIND_SPEED=min_max(WIND_SPEED),
                          VISIBILITY=min_max(VISIBILITY),
                          DEW_POINT_TEMPERATURE=min_max(DEW_POINT_TEMPERATURE),
                          SOLAR_RADIATION=min_max(SOLAR_RADIATION),
                          RAINFALL=min_max(RAINFALL),
                          SNOWFALL=min_max(SNOWFALL))                 
# `RENTED_BIKE_COUNT`, `TEMPERATURE`, `HUMIDITY`, `WIND_SPEED`, `VISIBILITY`, `DEW_POINT_TEMPERATURE`, `SOLAR_RADIATION`, `RAINFALL`, `SNOWFALL`

In [134]:
# Print the summary of the dataset again to make sure the numeric columns range between 0 and 1
summary(normalized_dummy)

  TEMPERATURE        HUMIDITY        WIND_SPEED       VISIBILITY    
 Min.   :0.0000   Min.   :0.0000   Min.   :0.0000   Min.   :0.0000  
 1st Qu.:0.3636   1st Qu.:0.4286   1st Qu.:0.1216   1st Qu.:0.4602  
 Median :0.5472   Median :0.5816   Median :0.2027   Median :0.8429  
 Mean   :0.5345   Mean   :0.5933   Mean   :0.2332   Mean   :0.7131  
 3rd Qu.:0.7080   3rd Qu.:0.7551   3rd Qu.:0.3108   3rd Qu.:1.0000  
 Max.   :1.0000   Max.   :1.0000   Max.   :1.0000   Max.   :1.0000  
 DEW_POINT_TEMPERATURE SOLAR_RADIATION       RAINFALL           SNOWFALL       
 Min.   :0.0000        Min.   :0.000000   Min.   :0.000000   Min.   :0.000000  
 1st Qu.:0.4412        1st Qu.:0.000000   1st Qu.:0.000000   1st Qu.:0.000000  
 Median :0.6107        Median :0.002841   Median :0.000000   Median :0.000000  
 Mean   :0.5977        Mean   :0.161326   Mean   :0.004261   Mean   :0.008828  
 3rd Qu.:0.7924        3rd Qu.:0.264205   3rd Qu.:0.000000   3rd Qu.:0.000000  
 Max.   :1.0000        Max.   :1.0000

In [135]:
# Save the dataset as `seoul_bike_sharing_converted_normalized.csv`
write_csv(normalized_dummy, "seoul_bike_sharing_converted_normalized.csv")

## Standardize the column names again for the new datasets


Since you have added many new indicator variables, you need to standardize their column names again by using the following code:


In [136]:
# Dataset list
dataset_list <- c('seoul_bike_sharing.csv', 'seoul_bike_sharing_converted.csv', 'seoul_bike_sharing_converted_normalized.csv')

for (dataset_name in dataset_list){
    # Read dataset
    dataset <- read_csv(dataset_name)
    # Standardized its columns:
    # Convert all columns names to uppercase
    names(dataset) <- toupper(names(dataset))
    # Replace any white space separators by underscore, using str_replace_all function
    names(dataset) <- str_replace_all(names(dataset), " ", "_")
    # Save the dataset back
    write.csv(dataset, dataset_name, row.names=FALSE)
}

Rows: 8465 Columns: 14
-- Column specification --------------------------------------------------------
Delimiter: ","
chr  (4): DATE, SEASONS, HOLIDAY, FUNCTIONING_DAY
dbl (10): RENTED_BIKE_COUNT, HOUR, TEMPERATURE, HUMIDITY, WIND_SPEED, VISIBI...

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 8465 Columns: 40
-- Column specification --------------------------------------------------------
Delimiter: ","
chr  (5): FUNCTIONING_DAY, Autumn, Spring, Summer, Winter
dbl (35): TEMPERATURE, HUMIDITY, WIND_SPEED, VISIBILITY, DEW_POINT_TEMPERATU...

i Use `spec()` to retrieve the full column specification for this data.
i Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 8465 Columns: 40
-- Column specification --------------------------------------------------------
Delimiter: ","
chr  (5): FUNCTIONING_DAY, Autumn, Spring, Summer, Winter
dbl (35

# Next Steps


Great! Now that you have processed all of the necessary datasets, you are ready to perform exploratory data analysis to get some inital insights from them.


## Authors

<a href="https://www.linkedin.com/in/yan-luo-96288783/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkRP0321ENSkillsNetwork25371262-2022-01-01" target="_blank">Yan Luo</a>


### Other Contributors

Jeff Grossman


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By | Change Description      |
| ----------------- | ------- | ---------- | ----------------------- |
| 2021-04-08        | 1.0     | Yan        | Initial version created |
|                   |         |            |                         |
|                   |         |            |                         |

## <h3 align="center"> © IBM Corporation 2021. All rights reserved. <h3/>
